File: benchmarks/runners/node/warm-bench.mjs
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { casesForMetric, resolveInputs } from "../../lib/manifest.mjs";
import { computeStats, toMBps } from "../../lib/stats.mjs";

const nowNs = () => process.hrtime.bigint();
const msSince = (start) => Number(nowNs() - start) / 1e6;

/** Build the wrapper Inputs map from resolved corpus inputs. */
function toInputsMap(resolved) {
  const inputs = {};
  for (const [name, v] of Object.entries(resolved)) {
    inputs[name] = { content: v.buffer, mimeType: v.mimeType, charset: v.charset ?? "utf-8" };
  }
  return inputs;
}

/** Split a Buffer into fixed-size chunks for streaming input. */
function* chunked(buffer, size = 65536) {
  for (let i = 0; i < buffer.length; i += size) yield buffer.subarray(i, i + size);
}

async function drain(gen) {
  let total = 0;
  let step = await gen.next();
  while (!step.done) {
    total += step.value.length;
    step = await gen.next();
  }
  if (!step.value.success) throw new Error(`stream failed: ${step.value.error}`);
  return total;
}

/**
 * @returns {Promise<Array<{id,metric,unit,stats,iterations}>>}
 */
export async function runWarmAndStreaming(api, manifest) {
  const rows = [];

  for (const c of casesForMetric(manifest, "warm")) {
    const script = readScript(manifest, c);
    const inputs = toInputsMap(resolveInputs(manifest, c));
    const warmup = c.iterations?.warmup ?? 10;
    const iters = c.iterations?.warm ?? 100;

    for (let i = 0; i < warmup; i++) assertOk(api.run(script, inputs));
    const samples = [];
    for (let i = 0; i < iters; i++) {
      const start = nowNs();
      assertOk(api.run(script, inputs));
      samples.push(msSince(start));
    }
    rows.push({ id: c.id, metric: "warm", unit: "ms", stats: computeStats(samples), iterations: iters });
  }

  for (const c of casesForMetric(manifest, "streaming")) {
    const script = readScript(manifest, c);
    const resolved = resolveInputs(manifest, c);
    const [primaryName, primary] = Object.entries(resolved)[0];
    const iters = c.iterations?.streaming ?? 10;

    const mbps = [];
    for (let i = 0; i < iters; i++) {
      const start = nowNs();
      const gen = api.runTransform(script, chunked(primary.buffer), {
        inputName: primaryName,
        mimeType: primary.mimeType,
        charset: primary.charset,
      });
      await drain(gen);
      mbps.push(toMBps(primary.buffer.length, msSince(start)));
    }
    rows.push({ id: c.id, metric: "streaming", unit: "MB/s", stats: computeStats(mbps), iterations: iters });
  }

  return rows;
}

function assertOk(result) {
  if (!result.success) throw new Error(`run failed: ${result.error}`);
  return result;
}

function readScript(manifest, c) {
  return readFileSync(join(manifest.corpusDir, c.script), "utf-8");
}

File: benchmarks/runners/python/warm_bench.py
"""In-process metrics: warm steady-state and streaming throughput. Timing mirrors
the Node runner: perf_counter_ns -> ms. NO JIT warmup floor (CPython has no JIT),
matching Node — unlike the engine's 2000-iter floor."""

import time

from manifest import cases_for_metric, read_script, resolve_inputs
from stats import compute_stats, to_mbps


def _to_inputs(resolved):
    inputs = {}
    for name, v in resolved.items():
        inputs[name] = {"content": v["bytes"], "mimeType": v["mimeType"],
                        "charset": v["charset"] or "utf-8"}
    return inputs


def _chunked(data, size=65536):
    for i in range(0, len(data), size):
        yield data[i:i + size]


def _assert_ok(result):
    if not result.success:
        raise RuntimeError(f"run failed: {result.error}")


def run_warm_and_streaming(api, manifest, warmup_cap=None, warm_cap=None, streaming_cap=None):
    rows = []

    for c in cases_for_metric(manifest, "warm"):
        script = read_script(manifest, c)
        inputs = _to_inputs(resolve_inputs(manifest, c))
        warmup = warmup_cap if warmup_cap is not None \
            else c.get("iterations", {}).get("warmup", 10)
        iters = warm_cap if warm_cap is not None \
            else c.get("iterations", {}).get("warm", 100)

        for _ in range(warmup):
            _assert_ok(api.run(script, inputs))
        samples = []
        for _ in range(iters):
            start = time.perf_counter_ns()
            _assert_ok(api.run(script, inputs))
            samples.append((time.perf_counter_ns() - start) / 1e6)
        rows.append({"id": c["id"], "metric": "warm", "unit": "ms",
                     "stats": compute_stats(samples), "iterations": iters})

    for c in cases_for_metric(manifest, "streaming"):
        script = read_script(manifest, c)
        resolved = resolve_inputs(manifest, c)
        primary_name, primary = next(iter(resolved.items()))
        iters = streaming_cap if streaming_cap is not None \
            else c.get("iterations", {}).get("streaming", 10)

        mbps = []
        for _ in range(iters):
            start = time.perf_counter_ns()
            stream = api.run_transform(
                script, _chunked(primary["bytes"]),
                input_name=primary_name, input_mime_type=primary["mimeType"],
                input_charset=primary["charset"],
            )
            for _chunk in stream:
                pass
            meta = stream.metadata
            if meta is None or not meta.success:
                raise RuntimeError(f"stream failed: {getattr(meta, 'error', 'no metadata')}")
            elapsed_ms = (time.perf_counter_ns() - start) / 1e6
            mbps.append(to_mbps(len(primary["bytes"]), elapsed_ms))
        rows.append({"id": c["id"], "metric": "streaming", "unit": "MB/s",
                     "stats": compute_stats(mbps), "iterations": iters})

    return rows
